### ISSUE TITLE
Latest: Translator vs. Challenger: Adversarial Agentic Learning for C-to-Rust Translation

### ISSUE THREAD

**Issue by jjakimoto** [jjakimoto/research-issues#1531]
**Topic:** Quantitative trading; machine learning for quant & trading; agentic coding techniques; LLMs & multimodal LLMs; and autonomous research (autoresearch / AI-scientist) systems — methods, papers, tools  
**Date:** 2026-09-15

**Source:** arxiv · 2026-09-14
**Link:** https://arxiv.org/abs/2609.15381

Instead of just accumulating 'lessons learned' in an agent memory, Trail treats each learned insight as a falsifiable hypothesis and has a Challenger agent write executable C/Rust counterexamples to break it — turning trace-specific experience into knowledge that transfers across benchmarks and lifting project-level test pass rate from 40% to 69% on CRUST-Bench.

## Executive summary

Trail is an adversarial agentic learning framework for project-level C-to-Rust translation built from three cooperating LLM agents: a Translator that mines candidate 'insights' from failures and accepted repairs, a Challenger that synthesizes executable counterexample challenges to falsify them, and a Refiner that converts compiler/test evidence into structured insight edits behind an evidence gate. Refinement runs at two levels — single-insight stress testing and compositional testing of co-activated insight sets — producing an insight bank retrieved at inference time to guide translation and execution-guided repair. On 100 CRUST-Bench projects with three backends, Trail beats the strongest LLM baseline by 23.1% relative syntax accuracy and 15.9% semantic accuracy (88% CompRate / 69% TestRate with gpt-5.4-mini vs SmartC2Rust's 66%/40%), and the insight bank transfers to a second benchmark without adaptation. Grounding is compilation plus tests only; fuzzing and differential testing are explicitly future work.

### Agent topology: Challenger / Translator / Refiner in a challenge-translate-refine loop

Trail is not a single translator agent. The Challenger synthesizes executable C-to-Rust challenges targeting the applicability boundary of an insight; the Translator solves the challenge with that insight injected as an active constraint; the Refiner attributes the resulting compiler/test evidence to a specific insight deficiency. An evidence gate replays the same challenge under the proposed update and accepts it only if the replay now succeeds. At inference time the topology changes to Translator / Repairer / Reflector.

- Each insight is a 4-tuple (Trigger, Goal, Constraint, Risk) plus semantic tags, stored as structured JSON, indexed in FAISS, embedded with BAAI/bge-base-en-v1.5.
- Failures are attributed to exactly one of four deficiency classes, each mapped to a field edit: Trigger gap, Goal mismatch, Constraint violation, Missing risk.
- Challenges pass three sanity checks before burning execution budget: Leak (de-dup), Activates (trigger actually fires), ToolCheck (C snippet runs, Rust harness type-checks, oracle passes under cargo test).
- Default of at most 3 adversarial refinement rounds per accepted insight; unsupported candidates are discarded rather than banked.

### Compositional adversarial learning — what distinguishes it from Reflexion/ExpeL-style memory

Individually-correct insights can conflict when co-activated on a shared call chain (a rule preserving a C NUL terminator versus one building a Rust slice that excludes it). Trail does counterfactual pairwise interaction analysis over insight fields and tags, labelling relations support / independent / conflict, then greedily builds compact co-activated insight sets maximizing a coverage-minus-conflict objective (max 3 insights per joint challenge). Joint challenges must force oracle dependence on the interaction — sanity checks reject cases where the insights are exercised independently.

- Predicted conflicts are flagged for executable testing, not rejected, because apparent conflicts often conflate internal C buffer representation with external Rust API obligations.
- Accepted compositional updates include field refinements, relation updates, co-activation preconditions, and precedence rules.
- A failure exposing a non-representable pattern spawns a new candidate insight, which must re-enter single-insight adversarial learning before banking.
- Support/conflict scores are LLM-estimated and used only as a heuristic prioritizer; real updates require compiler or test evidence.

### Verifier grounding: compilation + project tests, no fuzzing

Every acceptance decision bottoms out in cargo check and cargo test. Translation-time repair is budgeted: up to five compile-fix attempts per callable and three project-level test-repair attempts, with a repair accepted only if it succeeds or strictly reduces diagnostics/failing tests. The authors acknowledge in Threats to Validity that this does not establish semantic equivalence with the original C.

- Metrics: CompRate (cargo check), TestRate (all cargo test pass), SafeRate (% callables with no unsafe block), Idiomaticity (cargo clippy Lint Pass + Rustine Idiom Penalty).
- No differential fuzzing today — named as future work alongside stronger API specifications.
- Project structuring: macro expansion, whole-project dependency graph, topological translation order, then per-function insight retrieval (default top-5).

### Benchmarks and results

Primary benchmark is CRUST-Bench (100 real-world C projects from GitHub, Linux, PostgreSQL, each with an idiomatic Rust interface crate and executable tests); cross-benchmark transfer uses 20 of the 21 SmartC2Rust-Bench programs. Backends are gpt-5.4-mini, kimi-k2.5, and deepseek-v4-flash at temperature 1.0 and 128K max output tokens. Baselines span C2Rust (rule-based), C2SaferRust (hybrid), Direct Prompting, Self-Repair, and SmartC2Rust.

- gpt-5.4-mini: 88% CompRate / 69% TestRate, beating SmartC2Rust's CompRate by 22 points and Self-Repair's TestRate by 14 points.
- kimi-k2.5: 79%/67% (vs 76%/61% best baseline); deepseek-v4-flash: 74%/54% (vs 56%/48%).
- C2Rust and C2SaferRust hit 98%/98% but with 0.13% and 12.28% SafeRate and Idiom Penalty ~49 — raw pass rate alone is not translation quality.
- Transfer: the CRUST-Bench insight bank applied cold to SmartC2Rust-Bench lifts gpt-5.4-mini from 80%/55% to 85%/70%.

### Cost and operating envelope for practitioners

The adversarial loop is cheap on tokens but not on wall clock. Insight injection adds 9.8% tokens because most spend remains on translation, compiler feedback, and repair; the slowdown comes from trace reflection plus the challenge-translate-refine cycles. Sensitivity analysis says both the retrieval budget and the number of adversarial rounds saturate quickly.

- 21.43M tokens total (214.3K/project) vs 19.51M (195.1K/project) without insights — a 1.92M-token, 9.8% overhead.
- 15,326s total (153.3s/project) vs SmartC2Rust's 10,452s (104.5s/project), a 1.47x slowdown.
- Retrieval size k: k=1 gives 78%/56%, k=3 gives 84%/65%, default k=5 gives 88%/69%, larger budgets drop TestRate to 68% then 66%.
- Adversarial rounds: 0 rounds = 85%/63%, 1 = 86%/66%, 2 = 87%/68%, 3 = 88%/69%; beyond that only stabilizes.
- Artifact released at https://github.com/bbzswcf/TRAIL including all prompts.

**Table II (excerpt): Overall comparison on 100 CRUST-Bench projects**

| Model | Method | CompRate | TestRate | SafeRate | Lint Pass | Idiom Penalty |
| --- | --- | --- | --- | --- | --- | --- |
| — | C2Rust | 98% | 98% | 0.13% | 98 | 49.06 |
| gpt-5.4-mini | C2SaferRust | 98% | 98% | 12.28% | 94 | 49.20 |
| gpt-5.4-mini | Direct Prompting | 32% | 24% | 99.66% | 34 | 11.98 |
| gpt-5.4-mini | Self-Repair | 56% | 55% | 99.67% | 65 | 8.70 |
| gpt-5.4-mini | SmartC2Rust | 66% | 40% | 99.41% | 67 | 4.77 |
| gpt-5.4-mini | Trail (ours) | 88% | 69% | 96.63% | 88 | 3.95 |
| kimi-k2.5 | SmartC2Rust | 59% | 37% | 97.16% | 60 | 5.21 |
| kimi-k2.5 | Trail (ours) | 79% | 67% | 95.12% | 89 | 4.47 |
| deepseek-v4-flash | SmartC2Rust | 31% | 26% | 98.49% | 15 | 7.22 |
| deepseek-v4-flash | Trail (ours) | 74% | 54% | 97.57% | 88 | 4.22 |

**Table III (excerpt): Cumulative ablation (CompRate / TestRate)**

| Model | Variant | CompRate | TestRate | SafeRate |
| --- | --- | --- | --- | --- |
| gpt-5.4-mini | Trail (full) | 88% | 69% | 96.63% |
| gpt-5.4-mini | - Compositional adversarial learning | 86% | 66% | 96.22% |
| gpt-5.4-mini | - Single-insight adversarial learning | 85% | 63% | 96.60% |
| gpt-5.4-mini | - Insight mining | 75% | 56% | 93.87% |
| kimi-k2.5 | Trail (full) | 79% | 67% | 95.12% |
| kimi-k2.5 | - Insight mining | 70% | 49% | 94.98% |
| deepseek-v4-flash | Trail (full) | 74% | 54% | 97.57% |
| deepseek-v4-flash | - Insight mining | 68% | 45% | 96.85% |

**Table IV (excerpt): Cross-benchmark transfer to SmartC2Rust-Bench (20 programs)**

| Model | Setting | CompRate | TestRate | Lint Pass |
| --- | --- | --- | --- | --- |
| gpt-5.4-mini | Base (no insights) | 80% | 55% | 13/20 |
| gpt-5.4-mini | Direct (target-benchmark insights) | 90% | 75% | 16/20 |
| gpt-5.4-mini | Transferred (CRUST-Bench insights) | 85% | 70% | 13/20 |
| kimi-k2.5 | Base | 90% | 65% | 14/20 |
| kimi-k2.5 | Transferred | 90% | 70% | 15/20 |
| deepseek-v4-flash | Base | 75% | 70% | 16/20 |
| deepseek-v4-flash | Transferred | 80% | 70% | 17/20 |

## Key findings

- On 100 CRUST-Bench projects with gpt-5.4-mini, Trail reaches 88% CompRate and 69% TestRate, beating SmartC2Rust's 66% CompRate by 22 points and Self-Repair's 55% TestRate by 14 points; averaged across backends that is +23.1% relative syntax accuracy and +15.9% semantic accuracy.
- The single biggest contributor is having any mined insights at all: removing insight mining drops TestRate to 56% / 49% / 45% across the three backends, while the two adversarial-refinement layers add a further ~3-13 TestRate points on top.
- Unsafe-first transpilers win on raw pass rate and lose everywhere else: C2Rust hits 98%/98% but only 0.13% SafeRate and an Idiom Penalty of 49.06, versus Trail's 96.63% SafeRate and 3.95 penalty.
- Insights transfer cold across benchmarks (gpt-5.4-mini 80%/55% to 85%/70% on SmartC2Rust-Bench) but remain 5-10 points behind insights adversarially refined on the target benchmark.
- Both key hyperparameters saturate: retrieval of 5 insights is optimal (larger budgets cut TestRate from 69% to 66% via irrelevant constraints), and 2-3 adversarial rounds capture essentially all the gain.
- Overhead is modest on tokens (21.43M vs 19.51M, +9.8%) but real on latency (153.3s vs 104.5s per project, 1.47x slower than SmartC2Rust).

## Notable quotes

> Refinement is triggered only by actionable failures; translator noise, invalid challenges, and successful executions do not produce updates.
> If the replay succeeds, the update is accepted and becomes the insight for the next round.
> Removing insight mining yields the largest drop, with TestRate falling to 56%, 49%, and 45% on gpt-5.4-mini, kimi-k2.5, and deepseek-v4-flash.
> Future work will complement them with fuzzing, differential testing, and stronger API specifications.

## Limitations

- Correctness is grounded only in cargo check and cargo test, which the authors concede do not establish semantic equivalence with the original C; there is no fuzzing or differential testing in the verifier today.
- Compositional learning relies on LLM-estimated support/conflict scores that may drift across backends or decoding settings.
- Evaluation covers only public user-level C projects (100 + 20); kernel/driver code, highly concurrent systems and large industrial codebases are untested, and pre-training contamination on the original C sources cannot be ruled out.

## Open questions

- Would swapping the compile-and-test oracle for differential fuzzing against the original C binary change which insights survive the evidence gate — i.e. how many banked insights are only 'true' relative to a weak oracle?
- Does the insight bank keep improving with more projects, or does the greedy composition cap the reachable coverage of interacting C idioms?
- Can the Challenger/Refiner topology be lifted out of C-to-Rust into general agent experience memory, where the executable counterexample is the scarce ingredient?

---
_Generated by Auto Research via a headless coding agent. Sources were retrieved with live web search — please verify before citing._

**Comment by x711criptic** [comment]
Might be relevant here — x711 is a pay-per-use tool API for autonomous agents. 41 tools (web search, crypto prices, shared Hive memory, on-chain tx sim, email, agent spawning), free tier 10 calls/day:

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"web_search","query":"your query"}'
```

Works via REST or MCP. 1,000+ agents on it across LangChain, CrewAI, AutoGen, ElizaOS.

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by jjakimoto** [comment]
Superseded by a fresh Auto Research run — closing to reset the slate.
